README
kplex: A multiplexer for various nmea 0183 interfaces
Copyright Keith Young 2012-2020 <stripydog7@gmail.com>
Blurb:
Details of the GPLv3 under which this software is distributed can be found in
the file COPYING in this directory.

This software almost certainly contains bugs and should not be relied on as
part of a safety critical system, ie as a sole source of marine navigational
information.

Installation From Source
------------------------
You've got this far so you've unpacked it.  So "make" should be the obvious
next step.  If you're using FreeBSD or some other system where GNU make is not
default, you make have to type "gmake" instead.  There's only one executable
("kplex").

"make install" will install kplex into /usr/bin on Linux systems, /usr/local/bin
on other systems. You can change this by setting BINDIR. ie to install to
/usr/sw/bin use:
make -D BINDIR=/usr/sw/bin install
alternatively if using sh/ksh/bash:
BINDIR=/usr/sw/bin make install

The "install" target does not install a start script. Tihs must be done manually
if required. It does not create a configuration file.  The file "kplex.conf.ex"
can be used as a starting point to create your own configuration file using the
information contained in this README file.

"make uninstall" will remove the kplex binary.  If you specified a non-standard
installation location using BINDIR, specify it again for the uninstall target.

If you want to have kplex start on boot, kplex.init is an example init script
for debian-derived systems. It expects kplex to be installed in /usr/bin
and a configuration file in /etc/kplex.conf. Change these as
necessary, install by copying kplex.init to /etc/init.d/kplex and run:
update-rc.d kplex defaults
to create symlinks in the appropriate rc.d directories.

kplex.init will run kplex as root, but this is neither required nor recommended.
To automatically run kplex as a different user, change RUN_AS_USER as
appropriate.  The user kplex runs as does require permissions to read
/write serial interfaces as necessary.  Typically this will mean being a member
of the group which owns serial devices. On debian-based systems like ubuntu this
means adding the user to the "dialout" group.

From .deb file
--------------
If you've installed from a debian binary package, the startup script and binary
are installed for you. The default start script is not configured to run at
boot time but can be made to do so using update-rc.d(8). The start script runs
kplex as root but this is not recommended.  Instead create a user (e.g.
"kplex") and change the RUN_AS_USER definition in /etc/init.d/kplex to specify
that kplex run as that user. Be sure to add this user to the group which owns
the serial ports (normally "dialout" on modern debian-based systems).

The configuration file the .deb installation installs in /etc/kplex.conf
actually does nothing (all directives are commented out).  Read the instructions
in this README file for how to create a confiuration file.  A quick start
example which bridges between a tcp server and a 38400 baud serial-to-usb
connection is commented out in this example file.

Installation with Docker
------------------------
You can build the docker image by cloning the repository and running `docker build -t kplex .`.

Using Docker Compose
--------------------
Copy the `kplex.conf.ex` file to `kplex.conf` and define the default options as described in the next section. 

If you have configured a TCP/UDP port you need to open this port in `docker-compose.yml` by uncomment/modify the port section.

In case you are using a usb device you need to route the device to the container. Open the `docker-compose.yml`, uncomment and modify the volume pointing to `/dev`. You also need to uncomment `privileged: true` in ordeer for the container to access the `/dev` directory.

Invocation:
-----------

kplex [-V] | [-d 1..9] [-f <filename>] [-p <pidfile>] [-o <option>] <interface> <interface> [<interface> ...]

Where
The "-d" flag, if specified, will cause kplex to print debugging information,
the verbosity of which is controlled by the flag's argument with "1" producing
minimal additional debugging information and "9" producing verbose debug
information.

<filename> is the path of a configuration file to use where default options and
interfaces are defined. If not specified, kplex will look for its configuration
in the following locations and use the first file found:
* A location specified in the KPLEXCONF environment variable
* OSX ONLY: ~/Library/Preferences/kplex.ini.  This location is for backwards
  compatibility, will not be supported in the next major release and its use is
  discouraged.
* OS ONLY: ~/Library/Preferences/kplex.conf
* ~/.kplex.conf
* /etc/kplex.conf

Specifying filename as "-" disables looking for the default configuration files.

<pidfile> is the path of a file where kplex will write its pid for the
convenience of mechanisms such as init files used to control system processes.
If this option is specified and kplex finds that <pidfile> is locked by another
process it will print a warning and exit.

<option> is an option specifier for the form:
    <varable>=<value>
These correspond to configuration options which may otherwise be specified in
"global" section of the configuration file (discussed in the "Configuration
File" secction below).  Command line configuration options always override
options given in the configuration file.

<interface> is an interface specifier which takes the form:
<interface_type>:<option>[,<option> ...]
where:
    <interface_type> is one of:
        "serial": serial nmea 0183 data
        "file": file (or standard in/out)
        "tcp": nmea over TCP
        "udp": nmea over UDP (unicast/broadcast/multicast)
        "broadcast": nmea over UDP broadcast (DEPRECATED)
        "bcast": synonym for "broadcast" (DEPRECATED)
        "multicast": nmea over UDP multicast (DEPRECATED)
        "mcast": synonym for "multicast" (DEPRECATED)
        "pty": serial nmea data over a pseudo terminal
        "gofree": nmea over TCP announced by Navico's "GoFree" protocol
    <option> is one or more options of the form <var>=<val>. Some options are
        not optional. Options applicable to all interface types are:
        "direction":  May normally be one of "in" specifying an input, "out"
            specifying an output, or "both" specifying a bi-directional
            interface. Not all directions are applicable to all interface types.
            The default is "both" if an interface type permits bi-directional
            commnuication. It is more efficient to specify "in" or "out" if that
            is all an interface needs to do.
        "qsize": Size of the interface's output queue. Not used for input only
            interfaces.  Defaults should be fine. This should only need to be
            increased from default in the case of a bursty high-speed input
            feeding a slow ouput.
        "checksum": Overrides the global checksum option for an interface. If
            set to "no", no checksum checks are performed on incoming sentences.
            If set to "yes" or "strict", incoming sentences without correct
            checksums are discarded.  If set to "loose", sentences with
            incorrect checksums are discarded but those without any checksum
            are permitted. If set to "add", sentences with incorrect checksums
            are discarded but where no checksum is present, one will be added.
            If set to "addonly", a checksum will be added if absent but
            sentences with incorrect checksums are accepted.  With "add" and
            "addonly", sentences without checksums are discarded if the addition
            of a checksum would exceed the maximum permitted sentence length.
        "strict": May be "yes" to enable strict parsing of incoming sentences on
            an interface or "no" to disable it. This option overrides the global
            strict parsing option.
        "loopback": May be "yes" to enable sentences read from an interface to
            be written back to it when communication is bi-directional or "no"
            (the default) to ensure sentences are not looped back to the
            interface they were read from.  This option should be used with
            extreme caution and generally not at all with broadcast or
            multicast interfaces.  This option has no effect on unidirectional
            interfaces.
        "ifilter": Specifies an input filter (see below)
        "ofilter": Specifies an output filter (see below)
        "xifilter": Specifies an eXternal input filter (see below)
        "xofilter": Specifies an eXternal output filter (see below)
        "name": Attaches a symbolic name to an interface.  This is only required
            if you intend to use the interface for failover (see below) but can
            be helpful for debugging.  The value associated with "name" can be
            any string consisting of letters and or numbers which is not used as
            the "name" for another interface.  If no "name" is assigned kplex
            will autogenerate one starting with an underscore ('_').  For this
            reason names starting with an underscore should not be manually
            assigned.  Names are NOT case sensitive, so do not call one
            interface "Serial" and another "serial". 
        "srctag": Specifies that an NMEA-0183v4 TAG block containing a source
            identifier be prepended to sentences output on the interface if the
            value is "yes" or "input".  The src identification string is the
            interface name (see the "name" option above) truncated to 15
            characters or the string "kplex" if the interface name starts with a
            n underscore as is the case with autogenerated (non-user-specified)
            names. A TAG block prepended by the "srctag=yes" option looks like:
            \s:kplex*23\
            If "srctag=no" is specified, no source identifier is prepended.
            This is the default.
            If "srctag=input" is specified, the src identification string is the
            name of the interface on which the sentence arrived truncated to 15
            characters or the string "kplex" if no "name" is specified for the
            input interface.
        "timestamp": Specifies that an NMEA-0183v4 TAG block containing UNIX
            time (for example "\c:1423133110*5C\" should be prepended to
            sentences output on the interface.  The timestamp is in seconds if
            the value is "s" or milliseconds if the value is "ms".  Note that
            NMEA-0183v4 timestamps do not take account of leap seconds.
        "optional": If "optional=no" is specified or this option is not given,
            kplex will exit if it cannot initialize the interface. If
            "optional=yes" is specified, failure of the interface to initialize             will only cause kplex to exit if, as a result of it failing, kplex
            has no inputs or no outputs.
        "heartbeat": Specifies that kplex should send a proprietary sentence
            from the interface with a frequency specified as the option's value
            in seconds, i.e. "heartbeat=60" will send a sentence every minute.
            Other instances of kplex will not further propagate these
            "heartbeats". This option is not valid on input-only interfaces.

If source identifier and timestamps are both requested for an interface, they
are combined into a single TAG block, source identifier first, e.g.:
\s:kplex,c:1423133048*5F\

The -V flag instructs kplex to print its version number and exit. It should not
be used with any other options.

Interface Types:
----------------

Serial Interfaces
-----------------

This is the traditional way of getting nmea data into your computer.  kplex
doesn't care whether your device is connected to via a traditional serial
port or via a USB to serial converter so long as the device on which data are
presented looks like a character special device and it can be configured with
baud rate and a minimal set of other parameters. Note that you can't normally
just plug your nmea tx/rx into a serial port. NMEA is RS422 whereas serial ports
normally want RS232 input.

    Interface-specific options:
        filename=<device>
        baud=<baud>
        Where
            <device> is the serial device (e.g. /dev/ttyS0)
            <baud> is the baud rate.  Defaults to 4800 if unspecified
            Supported baud rates are: 4800, 9600, 19200, 38400, 57600, 115200,
            230400 and 460800.

You must minimally specify a device name for a serial interface.  usb to serial
converters often use /dev/ttyUSB0. Check your /var/adm/messages file and/or udev
rules.  Note that normal users are often not permitted to open serial devices.  This may mean adding your user to a group which *is* allowed to read the device
(e.g. "dialout", "uucp" or whatever).


"File" interfaces
-----------------

This specification covers both regular files, FIFOs and terminal i/o via
standard input and standard output.

    Interface-specific options:
        filename=<file>
        filenamex=<format>
        persist=[yes|no]
        append=[yes|no]
        eol=[rn|n]
        owner=<user>
        group=<group>
        perm=<permissions>
        Where
            <file> is either the file name to read from or write to or "-".
                In the latter case, standard input is used for inputs, standard
                out for outputs, and standard in and standard out for "BOTH".
                If not specified deafults to "-".
            <format> is either "-", which is treated as in the "file" case
                above, or a format string from which a filename is constructed.
                The format string substitues variables preceded by the "%"
                character per the strftime(3) library call.  Additionally,
                format strings in the format %{<keyword>} are substituted for
                the following values of <keyword>:
                "host": The hostname of the host on which kplex is running
            <user> is the user to set ownership of a created output file to.
            <group> is the group to set a created output file to.
            <permissions> are the file access permissions, in octal form, to set
                a created output file to.

"File" interfaces are slightly different from other interfaces in that
by default sentences are terminated by <LF> rather than <CR><LF>. Because this
is *nix and we don't want to muck about with redundant '\r's.  Uniquely for
"file" interfaces "loose" termination checking is default on input.  Sentences
input with just a terminating '\n' (ie <LF>) or NULL are converted to <CR><LF>
for output to other interface types (i.e. serial, tcp, broadcast
etc.).

If the option "eol=rn" is specified, file interfaces will output sentences

CHANGELOG
v1.5:
* i8n
* Added filenamex option for file interfaces
* Added heartbeat option
* Expanded checksumming options
* Added xfilter support
* Re-lookup up tcp server address on each reconnection
* Prevent bi-directional udp broadcast interfaces re-reading transmitted data

v1.4: January 6th 2019
* Fixed crash with checksumming enabled
* Various queue length changes
* Changed default input processing to "loose" for file interfaces
* Minor change to how default values for "strict" applied
* Fixed possible null pointer exception in serial interface argument processing
* Minor changes to persistent tcp behaviour
* Add support for B460800 and B230400 baud rates
* Make lowering file descriptor limits debug rather than warning
* Fix lack of null termination on interface names
* Make ~/.kplex.conf per-user conf file on OSX deprecating previous location
* Add -p option for pid file
* Add man page
* Fix errors in preamble parsing

v1.3.4: June 26th 2016
* Fixed bug with optional interfaces
* Fixed udp port assignment bug

v1.3.3: May 7th 2016
* Fixed crash in udp code with openvpn
* Fixed broken filtering

v1.3.2: April 21st 2016
* Added gpsd support
* Fixed various udp handling bugs
* Fixed various bugs on tcp reconnect
* Improve reconnection handling
* Added much more debugging
* Give interfaces default names

v1.3.1: September 27th 2015
* Fixed various uninitialised pointer fixes
* Updated Makefile for OpenWRT and github builds
* Accept NULL as a sentence terminator with "strict=no"
* Fixed problem with non-truncation of output files
* Better guessing of UDP interface parameters
* Define ACCESSPERMS, remove redundant declarations, add header guards
* Fixed bug in source filtering
* Added option to re-enable nagle for TCP interfaces
* "-d" flag now documented with additional debugging
* Default queue sizes reduced and SNDBUF size reduced
* Added optional interface to filter rules

v1.3: April 15th 2015
* Added "preamble" option to tcp interfaces
* Fixed bugs in failover specification parsing
* Added udp: interface type
* Added AIS coalescing
* Added strict parsing option
* Fixed bug in file output to new files

v1.2: January 6th 2015
* eol=n now optional on other interface types
* srctag=input option
* owner=,group= and perm= options for pty and output file interfaces
* Fixed bug in tcp server src TAGs
* Fixed multiple tcp persist bugs
* Remove pty symlinks on exit
* Updates for NetBSD portability
* Fixed TAG checksum bug

v1.1.1: November 12th 2014
* Added presist=fromstart option
* Fixed serial ports in OS X
* Fixed processing of file interfaces with "eol=rn"
* Fixed specification of multicast interface
* Fixed bug with outbound serial interface TAG processing
* Fixed bug with TAG timestamps
* Added fromstart value for persist option
* Minor comment corrections
* Minor changes for greater POSIX compatibility

v1.1: June 5th 2014
* Parsing algorithm changed to strip TAG blocks
* Added optional TAG-based timestamping
* Added optional TAG-based source identifiers
* Added GoFree support
* Added loopback option
* Added keepalive options to tcp interfaces
* Added timeout option to tcp interfaces
* Added more supported baud rates for serial and pty interfaces
* More consistent interface specifiers for broadcast and multicast
* Corrected bug in default port assignment for broadcast interfaces
* Fixed bugs in tcp persist code

v1.0: 31st August 2013
* Added multicast support
* Added "persist" and "retry" options for tcp client interfaces
* Added support for rate limiting
* Disable Nagle algorithm on tcp interfaces
* Fix bug in appending to non-existent files
* Fix hang when configuration with no inputs specified
* Corrected multiple typos in README

v0.2.11: 25th June 2013
* Fixed error with seatalk flag in serial.c
* Fixed makefile for bsd/macos
* Improved error reporting for broadcast interfaces
* Fix failure-to-exit bug when using tcp server
* Special handling of and improved support for FIFOs
* Ignore SIGPIPE to avoid crashes writing to FIFOs
* Added append mode for output files
* Enable use of redirected standard input and output in background mode
* Added persist flag for FIFOs
* Added eol option for \r\n line termination with file interfaces
* Make file output line buffered
* Reduce max sentence size to the 80 chars it was always meant to be
* Fixed tcp server filter inheritace bug
* Fixed bug resulting in qsize option being ignored for tcp and serial

v0.2.10: 31st March 2013
* Fixed crash when interface error condition encountered on startup

v0.2.9: 4th March 2013
* Fixed bug with input interfaces exiting before output startup
* Updated documentation to correctly describe use of "device" option
* Fixed segv in file error handling
* Fixed crash with input-only tcp servers
* Added "-V" flag for version identification

v0.2.8: 16th January 2013
* Fixed bug in shutdown of bi-directional tcp server
* Minor cosmetic source tweaks
* Minor non-code adjustments for Debian packaging
* Added example configuration kplex.conf.ex

v0.2.7: 27th December 2012
* Fixed crash and hang on exit bug with unidirectional tcp servers

v0.2.6: 18th December 2012
* Fixed "port out of range" problem with UDP broadcast interfaces

v0.2.5: 10th December 2012
* Fixed various POSIX compatibility isses

v0.2.4: 30th November 2012
* Support for FreeBSD
* Explicit enabling of support for IPv6-mapped-IPv4 for platforms where this is
  disabled by default

v0.2.3: 28th November 2012
* Fixed multiple problems with broadcast interfaces
* Fixed bug in handling of unclosed quotes arround arguments

v0.2.2: 27th November 2012
* Changed user config file to ~/Library/Preferences/kplex.ini on OS X
* Source tidy to silence compiler warnings on OS X

v.0.2: 21st November 2012
* Initial public test release


fileio.c
/* fileio.c
 * This file is part of kplex
 * Copyright Keith Young 2012 - 2020
 * For copying information see the file COPYING distributed with this software
 *
 * This file contains code for i/o from files (incl stdin/stdout)
 */

#include "kplex.h"
#include <stdlib.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/uio.h>
#include <pwd.h>
#include <grp.h>
#include <sys/utsname.h>
#include <sys/time.h>

#define KEYWORD_MAX 15
#define FNAME_MAX 255

struct if_file {
    int fd;
    char *filename;
};

/*
 * Duplicate struct if_file
 * Args: if_file to be duplicated
 * Returns: pointer to new if_file
 */
void *ifdup_file(void *iff)
{
    struct if_file  *newif;

    if ((newif = (struct if_file *) malloc(sizeof(struct if_file)))
        == (struct if_file *) NULL)
        return(NULL);

    memset ((void *)newif,0,sizeof(struct if_file));

    /* Read/Write only supported for stdin/stdout so don't allocate fp
     * And don't bother to duplicate filename
     */
    return(newif);
}

void cleanup_file(iface_t *ifa)
{
    struct if_file *iff = (struct if_file *) ifa->info;

    if (iff->fd >= 0)
        close(iff->fd);
    if (iff->filename)
        free(iff->filename);
}

void write_file(iface_t *ifa)
{
    struct if_file *ifc = (struct if_file *) ifa->info;
    senblk_t *sptr;
    int usereturn=flag_test(ifa,F_NOCR)?0:1;
    int data=0;
    int cnt=1;
    struct iovec iov[2];

    /* ifc->fd will only be < 0 if we're opening a FIFO.
     */
    if (ifc->fd < 0) {
        if ((ifc->fd=open(ifc->filename,O_WRONLY)) < 0) {
            logerr(errno,catgets(cat,4,1,
                    "Failed to open FIFO %s for writing\n"),ifc->filename);
            iface_thread_exit(errno);
        }
        DEBUG(3,catgets(cat,4,3,"%s opened FIFO %s for writing"),ifa->name,
                ifc->filename);
    }

    if (ifa->tagflags) {
        if ((iov[0].iov_base=malloc(TAGMAX)) == NULL) {
                logerr(errno,catgets(cat,4,4,"%s: Disabing tag output"),
                ifa->name);
            ifa->tagflags=0;
        } else {
            cnt=2;
            data=1;
        }
    }


    for(;;)  {
        if ((sptr = next_senblk(ifa->q)) == NULL) {
            break;
        }

        if (!usereturn) {
            sptr->data[sptr->len-2] = '\n';
            sptr->len--;
        }

        if (ifa->tagflags)
            if ((iov[0].iov_len = gettag(ifa,iov[0].iov_base,sptr)) == 0) {
                logerr(errno,catgets(cat,4,4,"%s: Disabing tag output"),
                        ifa->name);
                ifa->tagflags=0;
                cnt=1;
                data=0;
                free(iov[0].iov_base);
            }

        iov[data].iov_base=sptr->data;
        iov[data].iov_len=sptr->len;
        if (writev(ifc->fd,iov,cnt) <0) {
            if (!(flag_test(ifa,F_PERSIST) && errno == EPIPE) ) {
                logerr(errno,catgets(cat,4,5,"%s: write failed"),ifa->name);
                break;
            }

            if ((ifc->fd=open(ifc->filename,O_WRONLY)) < 0) {
                logerr(errno,catgets(cat,4,6,"%s: failed to re-open %s"),
                        ifa->name,ifc->filename);
                break;
            }
            DEBUG(4,catgets(cat,4,7,"%s: reconnected to FIFO %s"),ifa->name,
                    ifc->filename);
        }
        senblk_free(sptr,ifa->q);
    }

    if (cnt == 2)
        free(iov[0].iov_base);

    iface_thread_exit(errno);
}

void file_read_wrapper(iface_t *ifa)
{
    struct if_file *ifc = (struct if_file *) ifa->info;

    /* Create FILE stream here to allow for non-blocking opening FIFOs */
    if (ifc->fd == -1) {
        if ((ifc->fd = open(ifc->filename,O_RDONLY)) < 0) {
            logerr(errno,catgets(cat,4,8,
                    "Failed to open FIFO %s for reading\n"),ifc->filename);
            iface_thread_exit(errno);
        } else {
            DEBUG(3,catgets(cat,4,9,"%s: opened %s for reading"),ifa->name,
                    ifc->filename);
        }
    }
    do_read(ifa);
}

ssize_t read_file(void *ptr, char *buf)
{
    iface_t *ifa = (iface_t *)ptr;
    struct if_file *ifc = (struct if_file *) ifa->info;
    ssize_t nread;
    for(;;) {
        if ((nread=read(ifc->fd,buf,BUFSIZ)) <=0) {
            if (!flag_test(ifa,F_PERSIST))
                break;
            close(ifc->fd);
            if ((ifc->fd=open(ifc->filename,O_RDONLY)) < 0) {
                logerr(errno,catgets(cat,4,10,
                        "Failed to re-open FIFO %s for reading\n"),
                        ifc->filename);
                break;
            }
            DEBUG(4,catgets(cat,4,11,"%s: re-opened %s for reading"),ifa->name,
                    ifc->filename);
            continue;
        } else
            break;
    }
    return nread;
}

/*
 * replace a keyword with its expansion
 * Args: buffer to load the result to, size of buffer, keyword to expand
 * Returns: Number of characters in expanded buffer (excluding terminating
 * NULL) on success, 0 on failure or if expansion too large for buffer.
 */
size_t replace_keyword(char *buf, size_t max, const char * const keyword)
{
    struct utsname u;
    size_t n;

    if (!strcmp(keyword,"host")) {
        if (uname (&u) < 0) {
            return(0);
        }
        if ((n = strlen(u.nodename)) >= max) {
            return(0);
        }
        strcpy(buf,u.nodename);
        return(n);
    }

    return(0);
}

/*
 * Expand an extended filename string
 * Args: format string
 * Returns: expanded filename string on success, NULL on error
 */
char *expand_filename(const char * const format)
{
    char buf[FNAME_MAX+1],buf2[FNAME_MAX+1];
    char keyword[KEYWORD_MAX+1];
    const char *fptr;
    char *filename;
    int i,j,ret;
    int dotime=0;
    struct timeval tv;
    struct tm tms;

    errno = 0;

    for (i=0,fptr=format;i <= FNAME_MAX;) {
        if ((buf[i] = *fptr++) == '\0') {
            break;
        }
        if (buf[i] == '%') {
            if (*fptr == '{') {
                for (j=0,++fptr;j<KEYWORD_MAX;j++) {
                    if ((keyword[j] = *fptr++) == '}') {
                        keyword[j] = '\0';
                        break;
                    }
                }
                if (j == KEYWORD_MAX) {
                    return NULL;
                }
                if ((ret = replace_keyword(buf+i,sizeof(buf)-i,keyword)) == 0) {
                    return NULL;
                }

                i += ret;
                continue;
            } else {
                if (*fptr == '%') {
                    buf[++i] = *fptr++;
                }
                dotime=1;
            }
        }
        ++i;
    }
    if ( i > FNAME_MAX ) {
        return NULL;
    }

    ret = i;
    if (dotime) {
        (void) gettimeofday(&tv,NULL);
        localtime_r(&tv.tv_sec,&tms);
        if ((ret = strftime(buf2,sizeof(buf2),buf,&tms)) == 0) {
            return NULL;
        }
    }
    if ((filename = (char *) malloc(++ret)) == NULL) {
        return NULL;
    }
    strcpy(filename,(dotime)?buf2:buf);
    return filename;
}

iface_t *init_file (iface_t *ifa)
{
    struct if_file *ifc;
    struct kopts *opt;
    struct stat statbuf;
    int ret;
    int append=0;
    uid_t uid=-1;
    gid_t gid=-1;
    struct passwd *owner;
    struct group *group;
    mode_t tperm,perm=0;
    char *cp;

    if ((ifc = (struct if_file *)malloc(sizeof(struct if_file))) == NULL) {
        logerr(errno,catgets(cat,4,12,"Could not allocate memory"));
        return(NULL);
    }

    memset ((void *)ifc,0,sizeof(struct if_file));

    ifc->fd=-1;
    ifa->info = (void *) ifc;

    for(opt=ifa->options;opt;opt=opt->next) {
        if (!strcasecmp(opt->var,"filename")) {
            if (ifc->filename != NULL) {
                logerr(0,catgets(cat,4,37,
                        "Filename specified more than once"));
                return(NULL);
            }
            if (strcmp(opt->val,"-"))
                if ((ifc->filename=strdup(opt->val)) == NULL) {
                    logerr(errno,catgets(cat,4,13,
                            "Failed to duplicate argument string"));
                    return(NULL);
                }
        } else if (!strcasecmp(opt->var,"filenamex")) {
            if (ifc->filename != NULL) {
                logerr(0,catgets(cat,4,37,
                        "Filename specified more than once"));
                return(NULL);
            }
            if (strcmp(opt->val,"-"))
                if ((ifc->filename=expand_filename(opt->val)) == NULL) {
                    logerr(errno,catgets(cat,4,36,
                            "Failed to expand filenamex"));
                    return(NULL);
                }
        } else if (!strcasecmp(opt->var,"append")) {
            if (!strcasecmp(opt->val,"yes")) {
                append++;
            } else if (!strcasecmp(opt->val,"no")) {
                append = 0;
            } else {
                logerr(0,catgets(cat,4,15,"Invalid option \"append=%s\""),
                        opt->val);
                return(NULL);
            }
        } else if (!strcasecmp(opt->var,"owner")) {
            if ((owner=getpwnam(opt->val)) == NULL) {
                logerr(0,catgets(cat,4,16,"No such user '%s'"),opt->val);
                return(NULL);
            }
            uid=owner->pw_uid;
        } else if (!strcasecmp(opt->var,"group")) {
            if ((group=getgrnam(opt->val)) == NULL) {
                logerr(0,catgets(cat,4,17,"No such group '%s'"),opt->val);
                return(NULL);
            }
            gid=group->gr_gid;
        }
        else if (!strcasecmp(opt->var,"perm")) {
            for (cp=opt->val;*cp;cp++) {
                if (*cp >= '0' && *cp < '8') {
                    perm <<=3;
                    perm += (*cp-'0');
                } else {
                    perm = 0;
                    break;
                }
            }
            perm &= ACCESSPERMS;
            if (perm == 0) {
                logerr(0,catgets(cat,4,18,
                        "Invalid permissions for tty device \'%s\'"),opt->val);
                return 0;
            }
        } else {
            logerr(0,catgets(cat,4,19,"Unknown interface option %s\n"),
                    opt->var);
            return(NULL);
        }
    }

    /* We do allow use of stdin and stdout, but not if they're connected to
     * a terminal. This allows re-direction in background mode
     */
    if (ifc->filename == NULL) {
        if (flag_test(ifa,F_PERSIST)) {
            logerr(0,catgets(cat,4,20,
                    "Can't use persist mode with stdin/stdout"));
            return(NULL);
        }

        if (((ifa->direction != IN) &&
                (((struct if_engine *)ifa->lists->engine->info)->flags &
                K_NOSTDOUT)) ||
                ((ifa->direction != OUT) &&
                (((struct if_engine *)ifa->lists->engine->info)->flags &
                K_NOSTDIN))) {
            logerr(0,catgets(cat,4,21,
                    "Can't use terminal stdin/stdout in background mode"));
            return(NULL);
        }
        if (ifa->direction == IN) {
            ifc->fd = STDIN_FILENO;
            DEBUG(3,catgets(cat,4,22,"%s: using stdin"),ifa->name);
        } else {
            ifc->fd = STDOUT_FILENO;
            DEBUG(3,catgets(cat,4,23,"%s: using %s"),ifa->name,
                    (ifa->direction==OUT)?"stdout":"stdin/stdout");
        }
    } else {
        if (ifa->direction == BOTH) {
            logerr(0,catgets(cat,4,24,
                    "Bi-directional file I/O only supported for stdin/stdout"));
            return(NULL);
        }

        if ((ret=stat(ifc->filename,&statbuf)) < 0) {
            if (ifa->direction != OUT) {
                logerr(errno,"stat %s",ifc->filename);
                return(NULL);
            }
        }
        if ((ret == 0) && S_ISFIFO(statbuf.st_mode)) {
            /* Special rules for FIFOs. Opening here would hang for a reading
             * interface with no writer. Given that we're single threaded here,
             * that would be bad
             */
            if (access(ifc->filename,(ifa->direction==IN)?R_OK:W_OK) != 0) {
                logerr(errno,catgets(cat,4,25,"Could not access %s"),
                        ifc->filename);
                return(NULL);
            }
        } else {
            if (flag_test(ifa,F_PERSIST)) {
                logerr(0,catgets(cat,4,26,
                        "Can't use persist mode on %s: Not a FIFO"),
                        ifc->filename);
                return(NULL);
            }
            if (perm)
                tperm=umask(0);

            errno=0;
            /* If file is for output and doesn't currently exist...*/
            if (ifa->direction != IN && (ifc->fd=open(ifc->filename,
                        O_WRONLY|O_CREAT|O_EXCL|((append)?O_APPEND:0),
                        (perm)?perm:0664)) >= 0) {
                if (gid != 0 || uid != -1) {
                    if (chown(ifc->filename,uid,gid) < 0) {
                        logerr(errno,catgets(cat,4,27,
                                "Failed to set ownership or group on output file %s"),
                                ifc->filename);
                        return(NULL);
                    }
                }
            DEBUG(3,catgets(cat,4,28,"%s: created %s for output"),ifa->name,
                    ifc->filename);
            } else {
                if (errno && errno != EEXIST) {
                    logerr(errno,catgets(cat,4,29,"Failed to create file %s"),
                            ifc->filename);
                    return(NULL);
                }
                /* file is for input or already exists */
                if ((ifc->fd=open(ifc->filename,(ifa->direction==IN)?O_RDONLY:
                        (O_WRONLY|((append)?O_APPEND:O_TRUNC)))) < 0) {
                    logerr(errno,catgets(cat,4,30,"Failed to open file %s"),
                            ifc->filename);
                    return(NULL);
                }
                DEBUG(3,catgets(cat,4,31,"%s: opened %s for %s"),ifa->name,
                        ifc->filename,
                        (ifa->direction==IN)?catgets(cat,4,32,"input"):
                        catgets(cat,4,33,"output"));
            }
            /* reset umask: not really necessary */
            if (perm)
                (void) umask(tperm);
        }
    }

    free_options(ifa->options);

    ifa->write=write_file;
    ifa->read=file_read_wrapper;
    ifa->readbuf=read_file;
    ifa->cleanup=cleanup_file;

    if (ifa->direction != IN)
        if ((ifa->q = init_q(ifa->qsize,ifa->ofilter,ifa->name)) == NULL) {
            logerr(0,catgets(cat,4,34,"Could not create queue"));
            cleanup_file(ifa);
            return(NULL);
        }

    if (ifa->direction == BOTH) {
        if ((ifa->next=ifdup(ifa)) == NULL) {
            logerr(0,catgets(cat,4,35,"Interface duplication failed"));
            cleanup_file(ifa);
            return(NULL);
        }
        ifa->direction=OUT;
        ifa->pair->direction=IN;
        ifc = (struct if_file *) ifa->pair->info;
        ifc->fd=STDIN_FILENO;
    }
    return(ifa);
}
